### ISSUE TITLE
openshift.create() creates file on master in master/slave setup resulting in error

### ISSUE THREAD

**Issue by coreydaley** [openshift/jenkins-client-plugin#61]
In a master/slave scenario, when using `openshift.create()` within a `node('foo'){}` block, a temporary file named `/tmp/create<randomness>.markup` is created on the master, but the `oc create -f <tempfile>` is executed on the slave, resulting in an error: file not found.

Sample syntax that causes an issue:
```
openshift.withCluster() {
  openshift.withProject() {
    node('nodejs') {
      stage('create template') {
            git url: 'https://github.com/openshift/nodejs-ex.git' 
            def template = openshift.create(readFile('openshift/templates/nodejs-mongodb.json')).object()
      }
    }
  }
}
```

Sample error:
` err=error: the path "/tmp/create739820317386924410.markup" does not exist, verb=create, cmd=oc create -f /tmp/create739820317386924410.markup -o=name --server=https://172.30.0.1:443 --namespace=testing --token=XXXXX , out=, status=1}`

**Comment by gabemontero** [comment]
Marking as aa enhancement for now

@coreydaley and I talked about this one a little, and I had him even try a quick change of having the temp file created off of the run's workspace dir vs. the temp dir, but there was no change in result; also talked with @jupierce and he confirmed that when he developed the feature he only attempted master-only configurations.

 ... this is going to boil down to research into jenkins to see about if it is possible to access dynamically created files from the slave

if we can, then we'll move to make that change under the guise of an enhancement

if we cannot, then we'll move to fixing the documentation (with perhaps some prints within `openshift.create`) to clarify that the option on works on master

**Comment by jupierce** [comment]
@gabemontero @coreydaley - Can we try a couple things:
1. With the sleep step @coreydaley used to check if the temp file was *not* on the slave, can you verify it is on the master with the existing script?
2. Move the node() { } to the outside of the withCluster() {} and try the script again. Looking at the code, the DSL context is being established on the master.

**Comment by coreydaley** [comment]
I can confirm that the file was created on the master.

**Comment by coreydaley** [comment]
I can also confirm that moving the node(){} outside of the withCluster(){} worked, it created the file on the slave node and the script worked fine.

**Comment by jupierce** [comment]
@gabemontero Based on https://github.com/openshift/jenkins-client-plugin/issues/61#issuecomment-337615629 , I don't consider this a bug.  Maybe just a doc clarification necessary?
 
`openshift.withCluster( ... )` outside of node{} is equivalent to
```
node {
    openshift.withCluster(...) { 
    }
}
```

So @coreydaley's code, expanded out to:
```
node {   // this effectively selected the master
    openshift.withCluster(...) { 
      ....
         node('nodejs') {
              stage('create template') {
         }
    }
}
```

So, if a specific node is desired, it should surround withCluster. Best practice might to be to always encourage node{ withCluster{} }

**Comment by gabemontero** [comment]
@jupierce 
+1 not bug
+1 docs clarification, but with perhaps a printed warning when using `openshift.create(fileref)` as well ...

**Comment by jupierce** [comment]
@gabemontero It may also be possible to stream the input into stdin of create/apply to bypass the need to create the file. That would eliminate the potential for this error.

**Comment by gabemontero** [comment]
On Thu, Oct 19, 2017 at 11:09 AM, Justin Pierce <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> It may also be possible to
> stream the input into stdin of create/apply to bypass the need to create
> the file. That would eliminate the potential for this error.
>
Ah interesting ... I'll probably do the doc change short term, then give
your idea a go subsequently - thanks @jupierce

**Comment by gabemontero** [comment]
@coreydaley - based on the other 5 items I assigned you goes, let's level set when you wrap those up and see where each of us are with our to-do's this sprint.  Based on the results, we can see if we want one of use to tackle this guy.

Minimally the doc update per https://github.com/openshift/jenkins-client-plugin/issues/61#issuecomment-337750898
Possibly the refactor per https://github.com/openshift/jenkins-client-plugin/issues/61#issuecomment-337938634

**Comment by gabemontero** [comment]
Some issue cleanup ... 

The onslaught of declarative pipeline effectively forced us to update the docs and cite the best practice of including all our global variable invocations within a `script {...}`, which per pipeline conventions is always the inner most closure wrt the pipeline defined closures.

Going to punt on further muddying the waters with a special caveat for the scripted syntax `node {`, given the lack of external discovery of this since we opened this (ignoring the scripted vs. declarative difference).

On the how to refactor front to circumvent, in investigating the recent issues running `withCluster` inside of `parallel { .. }` blocks, cross ref'ed recent versions of docker workflow plugin and passing context into closures via env vars suffxed by job/run name via `withEnv`.  Updated https://trello.com/c/p4XCLN0A/1469-5-rethink-behavior-of-withcluster-etc-jenkinsintegration
